widgets: add onclick feature

Memoraike opened the pull request:
Implement feature requested in https://github.com/hyprwm/hyprlock/issues/366.

PaideiaDilemma commented on src/core/hyprlock.cpp:
```diff
@@ -682,6 +682,20 @@ void CHyprlock::handleKeySym(xkb_keysym_t sym, bool composed) {
+        const auto widgets = g_pRenderer->getOrCreateWidgetsFor(o->sessionLockSurface.get());
+        for (const auto& widget : *widgets) {
+            if (widget->containsPoint(pos)) {
+                widget->onClick(button, down, pos);
```
no `{}`

PaideiaDilemma commented on src/renderer/widgets/Label.cpp:
```diff
@@ -156,3 +157,25 @@ void CLabel::renderUpdate() {
     g_pHyprlock->renderOutput(outputStringPort);
 }
+
+bool CLabel::containsPoint(const Vector2D& pos) const {
+    if (!asset)
+        return false;
+
+    Vector2D mPos = {pos.x, abs(pos.y - viewport.y)};
+
+    CBox     box = {this->pos.x, this->pos.y, asset->texture.m_vSize.x, asset->texture.m_vSize.y};
+    box.rot      = angle;
+
+    bool isInside = box.containsPoint(mPos);
+
+    return isInside;
+}
```
Can you make another virtual function for IWidget that's is called smth like `getBoundingBox` and then move contains point to IWidget.cpp?

PaideiaDilemma commented on src/renderer/widgets/Label.cpp:
```diff
@@ -156,3 +157,25 @@ void CLabel::renderUpdate() {
+
+void CLabel::onClick(uint32_t button, bool down, const Vector2D& pos) {
+    if (down) {
+        if (!onclickCommand.empty()) {
```
no `{}`

PaideiaDilemma reviewed (commented):
Generally looks good.
It would be even better if we had something like `CLICKABLE` analogous to `SHADOWABLE`, to easily add that to all widgets.
But we could also add that in a next step.

vaxerski commented:
future things to consider:
 - hover
 - pointer

Memoraike commented:
Moved 'containsPoint' to 'IWidget', added a virtual method 'getBoundingBox'.
Added a virtual 'onHover' method and a 'hovered' state to 'IWidget'. For example, I added a cursor change for the 'CLabel' and 'CPasswordInputField' widgets on hover.
Added a 'CLICKABLE' macro for future expansion.

PaideiaDilemma commented on src/core/hyprlock.cpp:
```diff
@@ -682,6 +682,58 @@ void CHyprlock::handleKeySym(xkb_keysym_t sym, bool composed) {
+                    outputNeedsRedraw = true;
+                }
+                
+                if (!cursorChanged) {
```
single if no `{}`

PaideiaDilemma commented on src/renderer/widgets/Label.hpp:
```diff
@@ -22,6 +22,9 @@ class CLabel : public IWidget {
     void         renderUpdate();
     void         onTimerUpdate();
     void         plantTimer();
+    CBox         getBoundingBox() const override;
```
style: just virtual, no override.

PaideiaDilemma commented on src/renderer/widgets/IWidget.hpp:
```diff
@@ -17,6 +18,13 @@ class IWidget {
     static int      roundingForBox(const CBox& box, int roundingConfig);
     static int      roundingForBorderBox(const CBox& borderBox, int roundingConfig, int thickness);
 
+    virtual CBox    getBoundingBox() const {
```
Nit: maybe add a comment, or rename it to make it clear that this has to return Top-left position and size.

Memoraike replied:
What about `getBoundingClientRect`?

PaideiaDilemma replied:
`posFromHVAlign` returns bottom-left positions, while wayland typically uses top-left. So just `getBoundingBoxWL` would work. `getBoundingClientRect` is also fine :)

PaideiaDilemma commented:
works nicely!

PaideiaDilemma commented:
Hey @Memoraike,
now that 0.7.0 is released, let's get the fresh stuff in.
Can you fix the conflict and the unresolved convo?

Memoraike commented:
Hey, @PaideiaDilemma! Of course, I'll get on it starting Monday!

JunaidQrysh commented:
@Memoraike any updates on this? I can't wait for this banger MR.

PaideiaDilemma commented:
@Memoraike let me know if you want me to rebase it. I did change some widget stuff.

rotmh commented:
> @Memoraike any updates on this? I can't wait for this banger MR.

Could not agree more! I recently encountered this PR; What's the status of it?

PaideiaDilemma commented:
Succeeded by https://github.com/hyprwm/hyprlock/pull/736
